File: python/dnd-character/dnd_character.py
from math import floor
import random

class Character:
    def __init__(self):
        random.seed()
        for i in ['strength', 'dexterity', 'constitution', 'intelligence', 'wisdom', 'charisma']:
            exec(f'self.{i}={self.ability()}')
        self.hitpoints = 10 + modifier(self.constitution)

    
    def ability(self):
        scores = [random.randrange(1,7) for _ in range(4)]
        scores.remove(min(scores))
        return sum(scores)


def modifier(constitution):
    return floor((constitution - 10) / 2)

File: typescript/dnd-character/dnd-character.ts
export class DnDCharacter {
  strength: number;
  dexterity: number;
  constitution: number;
  intelligence: number;
  wisdom: number;
  charisma: number;
  hitpoints: number;

  constructor() {
    this.strength = DnDCharacter.generateAbilityScore();
    this.dexterity = DnDCharacter.generateAbilityScore();
    this.constitution = DnDCharacter.generateAbilityScore();
    this.intelligence = DnDCharacter.generateAbilityScore();
    this.wisdom = DnDCharacter.generateAbilityScore();
    this.charisma = DnDCharacter.generateAbilityScore();
    this.hitpoints = 10 + DnDCharacter.getModifierFor(this.constitution);
  }

  public static generateAbilityScore(): number {
    let arr: number[] = [];
    Array.from(Array(4)).forEach(() => {
      arr.push(this.getRandomIntInclusive(1, 7));
    });
    return arr.filter((_, i) => i).reduce((a, b) => a + b, 0);
  }

  public static getModifierFor(abilityValue: number): number {
    return Math.floor((abilityValue - 10) / 2);
  }

  public static getRandomIntInclusive(min: number, max: number): number {
    min = Math.ceil(min);
    max = Math.floor(max);
    return Math.floor(Math.random() * (max - min + 1) + min); //The maximum is inclusive and the minimum is inclusive
  }
}
